src/x86/runtime.rs

```diff
@@ -111,48 +111,56 @@ fn inv_test_bit(v: usize, idx: u32) -> bool {
 
-    // CPUID call with EAX=7, ECX=0 => Extended Features in EBX and ECX (unneeded):
-    if inv_test_bit(ebx, 3) { value = set_bit(value, __Feature::bmi as u32); }
-    if inv_test_bit(ebx, 5) { value = set_bit(value, __Feature::avx2 as u32); }
-    if inv_test_bit(ebx, 8) { value = set_bit(value, __Feature::bmi2 as u32); }
-
-    // CPUID call with EAX=1 => feature bits in ECX and EDX:
-    if inv_test_bit(ecx, 0) { value = set_bit(value, __Feature::sse3 as u32); }
-    if inv_test_bit(ecx, 5) { value = set_bit(value, __Feature::abm as u32); }
-    if inv_test_bit(ecx, 9) { value = set_bit(value, __Feature::ssse3 as u32); }
-    if inv_test_bit(ecx, 12) { value = set_bit(value, __Feature::fma as u32); }
-    if inv_test_bit(ecx, 19) { value = set_bit(value, __Feature::sse4_1 as u32); }
-    if inv_test_bit(ecx, 20) { value = set_bit(value, __Feature::sse4_2 as u32); }
-    if inv_test_bit(ecx, 21) { value = set_bit(value, __Feature::tbm as u32); }
-    if inv_test_bit(ecx, 23) { value = set_bit(value, __Feature::popcnt as u32); }
-    if inv_test_bit(ecx, 28) { value = set_bit(value, __Feature::avx as u32); }
-
-    if inv_test_bit(edx, 25) { value = set_bit(value, __Feature::sse as u32); }
-    if inv_test_bit(edx, 26) { value = set_bit(value, __Feature::sse2 as u32); }
+    if inv_test_bit(extended_features_ebx, 3) { value = set_bit(value, __Feature::bmi as u32); }
+    if inv_test_bit(extended_features_ebx, 8) { value = set_bit(value, __Feature::bmi2 as u32); }
+
+    if inv_test_bit(proc_info_ecx, 0) { value = set_bit(value, __Feature::sse3 as u32); }
+    if inv_test_bit(proc_info_ecx, 5) { value = set_bit(value, __Feature::abm as u32); }
+    if inv_test_bit(proc_info_ecx, 9) { value = set_bit(value, __Feature::ssse3 as u32); }
+    if inv_test_bit(proc_info_ecx, 12) { value = set_bit(value, __Feature::fma as u32); }
+    if inv_test_bit(proc_info_ecx, 19) { value = set_bit(value, __Feature::sse4_1 as u32); }
+    if inv_test_bit(proc_info_ecx, 20) { value = set_bit(value, __Feature::sse4_2 as u32); }
+    if inv_test_bit(proc_info_ecx, 21) { value = set_bit(value, __Feature::tbm as u32); }
+    if inv_test_bit(proc_info_ecx, 23) { value = set_bit(value, __Feature::popcnt as u32); }
+
+    if inv_test_bit(proc_info_edx, 25) { value = set_bit(value, __Feature::sse as u32); }
+    if inv_test_bit(proc_info_edx, 26) { value = set_bit(value, __Feature::sse2 as u32); }
+
+    // ECX[26] detects XSAVE and ECX[27] detects OSXSAVE, that is, whether the
+    // OS is AVX enabled and supports saving the state of the AVX/AVX2 vector
+    // registers on context-switches, see:
+    //
+    // - https://software.intel.com/en-us/blogs/2011/04/14/is-avx-enabled
+    // - https://hg.mozilla.org/mozilla-central/file/64bab5cbb9b6/mozglue/build/SSE.cpp#l190
+    //
+    if inv_test_bit(proc_info_ecx, 26) && inv_test_bit(proc_info_ecx, 27) {
+        if inv_test_bit(proc_info_ecx, 28) { value = set_bit(value, __Feature::avx as u32); }
```

Comment:
You need to actually use the xgetbv instruction here to confirm that the OS is going to do the right thing.